### PR TITLE
feat(parsers.avro): Support multiple modes for union handling

### LINKED ISSUE #12970: Avro processor doesnt support Unions
### Use Case

Telegraf Avro processor doesnt support unions, here is the sample schema and encoded string.

```
schemaJSON := `
	{
		"type": "record",
		"name": "test",
		"fields": [
			{"name": "field1", "type": "int"},
			{"name": "field2", "type": ["null", "string"]}
		]
	}`
```

// Example Avro message with a union type
	messageBytes := []byte("\x02\x0Ahello")

### Expected behavior

Avro processor to support unions

### Actual behavior

Avro processor failed with error " Could not convert map[string:abc] to string for tag "field2": type "map[string]interface {}" unsupported".

https://github.com/influxdata/telegraf/blob/master/plugins/parsers/avro/parser.go#L143

the conversion should handle type map[string]interface{}.

### Additional info

https://github.com/influxdata/telegraf/pull/11816

### PR DESCRIPTION
**Pull request by athornton**
# Required for all PRs

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #12970

Added a config item to allow user to specify "flatten" (current behavior), "nullable" (what the issue requestor wanted, and what we at Rubin Observatory need internally), and "any" (just take what comes in and put it in the named field) as modes by which Avro can handle unions.

The common use case of "nullable"--a field that can sometimes be null/missing, and if so, we get a measurement without that metric--is the major enhancement.

Either "nullable" with multiple non-null types or "any" would make InfluxDB unhappy with the received data (the entire measurement will be dropped when InfluxDB sees a metric of a type other than whatever it first saw (for a given shard)), but if getting data into InfluxDB isn't your goal, those modes could be useful.

### CODE REVIEW

**Comment by athornton** [comment]
Once #13939 is merged I'll rebase.

**Comment by athornton** [comment]
I have a question that comes from @afausti -- what should the default behavior be if you explicitly mark an incoming field as both a tag and a field?  The current code "does both" which ... I don't know what that does in practice.  The right answer is "don't do that," of course...but should I catch it when building the parser and warn and leave it a tag but not add it as a field?

File: plugins/parsers/avro/README.md [hunk 1]
```diff
@@ -88,6 +88,30 @@ The message is supposed to be encoded as follows:
...
+  ## not change the field name and will just put in whatever value it
+  ## receives.
+  ## WARNING: if you use "nullable" with more than one non-null type, or
+  ## if you use "any", and Telegraf is feeding InfluxDB, InfluxDB will
+  ## associate that field with the first type it sees for a given
+  ## its value.  If it receives another measurement with a different
+  ## type in that field, it will discard that entire measurement.  Be
+  ## sure you know what you're doing if you use the "any" type, or
+  ## "nullable" with more than one non-null type.
+  # avro_union_mode = "flatten"
```
Comment (srebhan) [inline plugins/parsers/avro/README.md:113]: Please move the explanation out of the config into a readme markdown section. I would shorten this to
```suggestion
  ## Define handling of union types. Possible values are:
  ##   flatten  -- add type suffix to field name (default)
  ##   nullable -- do not modify field name but discard "null" field values
  ##   any      -- do not modify field name and set field value to the received type
  # avro_union_mode = "flatten"
```

File: plugins/parsers/avro/parser.go [hunk 2]
```diff
@@ -50,6 +51,14 @@ func (p *Parser) Init() error {
 	default:
 		return fmt.Errorf("unknown 'avro_format' %q", p.Format)
 	}
+	switch p.UnionMode {
+	case "":
+		p.UnionMode = "flatten"
+	case "flatten", "nullable", "any":
+		// Do nothing as those are valid settings
+	default:
+		return fmt.Errorf("unknown 'avro_union_mode' %q", p.Format)
```
Comment (srebhan) [inline plugins/parsers/avro/parser.go:60]:
```suggestion
		return fmt.Errorf("unknown avro_union_mode %q", p.Format)
```

**Review by srebhan — COMMENTED** [review]
Thanks for tackling this @athornton! Just two small comments...

**Comment by athornton** [comment]
Review suggestions applied, and rebased onto current master.

**Review by srebhan — APPROVED** [review]
Thanks for the update @athornton!

File: plugins/parsers/avro/testdata/union-array/expected.out [hunk 3]
```diff
@@ -0,0 +1 @@
+array,name=pi data_0=3,data_1=3.0999999046325684,data_2=3.140000104904175,data_3=3.1410000324249268 1682509200092000
```
Comment (powersj) [inline plugins/parsers/avro/testdata/union-array/expected.out:1]: is the float parsing happening in the upstream library? Is there a reason for the loss?

File: plugins/parsers/avro/testdata/union-array/telegraf.conf [hunk 4]
```diff
@@ -0,0 +1,25 @@
+[[ inputs.file ]]
+  files = ["./testdata/json-array/message.json"]
```
Comment (powersj) [inline plugins/parsers/avro/testdata/union-array/telegraf.conf:2]: This is the wrong message.json ;)

File: plugins/parsers/avro/testdata/union-array/expected.out (hunk 3 above)
Comment (athornton) [reply plugins/parsers/avro/testdata/union-array/expected.out:1]: Well, the reason for the loss is that we're parsing it as a 32-bit float (because that's what the schema says), but rendering it into the metric with 15 digits of claimed precision because the InfluxDB Float is inherently 64-bit...so the bottom 32 bits are all zero, and that is not exactly the same as the nice decimal value (which also isn't exactly the same, floating point numbers aren't real numbers, etc., etc.).

Go Playground gives the same results:

```
package main

import (
	"fmt"
	"strconv"
)

func main() {
	vals := []string{"3", "3.1", "3.14", "3.141"}
	for _, v := range vals {
		f, err := strconv.ParseFloat(v, 32)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
		}
		fmt.Printf("string %s -> float %.15f\n", v, f)
	}
}
```

I can make the test suite cleaner by making the input values "double" if you want, which would give us the correct answer to 15 digits, but it is in fact the case that if you're sending 32-bit floats (which are a legal Avro data type), you're going to have some imprecision once it's in line protocol format.

This is, I think, correct (but surprising) float-conversion behavior.

Comment (powersj) [reply plugins/parsers/avro/testdata/union-array/expected.out:1]: > This is, I think, correct (but surprising) float-conversion behavior.

Yep, you are right, I missed the 32-bit.

File: plugins/parsers/avro/testdata/union-array/telegraf.conf (hunk 4 above)
Comment (powersj) [reply plugins/parsers/avro/testdata/union-array/telegraf.conf:2]: Did you push the fix for this? Ready to hit merge ;)

Comment (athornton) [reply plugins/parsers/avro/testdata/union-array/telegraf.conf]: Yeah, but not until I had chased down the 32-64 bit stuff.  Should be good now.

**Review by powersj — APPROVED** [review]
